### PR TITLE
Integration of CCPP-ized TJ2016 implementation

### PR DESCRIPTION
**Pull request by mwaxmonsky**
Updates atmospheric_physics to the latest tag bringing in TJ2016 capability and additional changes needed in support of TJ2016.

Describe any changes made to build system: N/A

Describe any changes made to the namelist: N/A

List any changes to the defaults for the boundary datasets: N/A

Describe any substantial timing or memory changes: N/A

Code reviewed by: nusbaume, peverwhee

List all existing files that have been modified, old files eliminated, new files added and describe the changes:

M       Externals_CAM.cfg
- Updating to latest CCPP Framework needed for build.
- Updating to latest atmospheric_physics needed to bring in TJ2016.

M       src/control/cam_comp.F90
- Removing CAM suite and dynamic_constituents to match new API.

M       src/data/physconst.meta
- Updating two standard names.

M       src/data/registry.xml
A       src/dynamics/utils/hycoef.meta
- Exposing sum_of_sigma_pressure_hybrid_coordinate_a_coefficient_and_sigma_pressure_hybrid_coordinate_b_coefficient (etamid) to physics schemes.

M       src/dynamics/utils/hycoef.F90
- Updating lev, ilev, hybi, hyai to be read in by their standard names and updating PIO checks appropriately.

A       tools/generate_input_to_stdnames_update.py
- Adding new script to convert CSV file exported from standard names spreadsheet into the inputnames_dictionary.xml

M       tools/inputnames_to_stdnames.py
- Updating script to be tphysac/tphysbc aware to avoid duplicate standard names in snapshots.

M       tools/stdnames_to_inputnames_dictionary.xml
- Adding missing entry for standard_gravitational_acceleration.

M       cime_config/testdefs/testlist_cam.xml
D       cime_config/testdefs/testmods_dirs/cam/outfrq_held_suarez_izumi_nooutput/shell_commands
D       cime_config/testdefs/testmods_dirs/cam/outfrq_held_suarez_izumi_nooutput/user_nl_cam
D       cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_izumi_nooutput/shell_commands
D       cime_config/testdefs/testmods_dirs/cam/outfrq_kessler_izumi_nooutput/user_nl_cam
A       cime_config/testdefs/testmods_dirs/cam/outfrq_tj2016_derecho_nooutput/shell_commands
A       cime_config/testdefs/testmods_dirs/cam/outfrq_tj2016_derecho_nooutput/user_nl_cam
- Adding derecho build test for tj2016, updating existing tests to build using both compilers (gnu and intel) and removing izumi tests based on discussion from 7/15/24 to simplify test setup process.

If there were any failures reported from running test_driver.sh on any test
platform, and checkin with these failures has been OK'd by the gatekeeper,
then copy the lines from the td.*.status files for the failed tests to the
appropriate machine below.  All failed tests must be justified.

derecho/intel/aux_cam
derecho/gnu/aux_cam
- All tests pass.

Summarize any changes to answers, i.e.,
- what code configurations: N/A
- what platforms/compilers: N/A
- nature of change (roundoff; larger than roundoff but same climate; new climate): N/A

### CODE REVIEW

**Review by nusbaume — CHANGES_REQUESTED** [review]
Looks good! I know it appears like a decent number of change requests but most of them should just be related to changing a particular error message.

Also, can the PR description be updated to include the following two things:

1.  The PR URL (so we can grab it in the git log if need be, unless Github will do this for us?)

2.  The results of the regression tests run on Derecho.

Instructions for how to run the tests can be found here:

https://github.com/ESCOMP/CAM-SIMA/wiki/CAM-SIMA-testing#regression-testing-manual

Note that you'll need to do it twice, once for GNU and once for Intel.  It might also be good to do snapshot test runs with the "official" snapshots that will live on Derecho here:

`/glade/campaign/cesm/community/amwg/sima_baselines/cam_sima_test_snapshots`

Mostly we just want to have a record that everything passed.

Thanks, and please let me know if you have any questions or concerns with any of my suggestions!

File: src/dynamics/utils/hycoef.F90 [hunk 1]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
+   ! Set PIO to return error codes.
+   call pio_seterrorhandling(file, PIO_BCAST_ERROR, pio_errtype)
+
    ! PIO traps errors internally, no need to check ierr
 
    ierr = PIO_Inq_DimID(File, 'lev', lev_dimid)
+   if (ierr /= PIO_NOERR) then
+      ierr = PIO_Inq_DimID(File, 'reference_pressure_in_atmosphere_layer', lev_dimid)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': reading lev')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90]: Updated!

File: src/dynamics/utils/hycoef.F90 [hunk 2]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
    if (pver /= flev) then
       write(iulog,*) routine//': ERROR: file lev does not match model. lev (file, model):',flev, pver
       call endrun(routine//': ERROR: file lev does not match model.')
    end if
 
    ierr = PIO_Inq_DimID(File, 'ilev', lev_dimid)
+   if (ierr /= PIO_NOERR) then
+      ierr = PIO_Inq_DimID(File, 'reference_pressure_in_atmosphere_layer_at_interfaces', lev_dimid)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': reading ilev')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90]: Updated!

File: src/dynamics/utils/hycoef.F90 [hunk 3]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
    if (pverp /= filev) then
       write(iulog,*) routine//':ERROR: file ilev does not match model ilev (file, model):',filev, pverp
       call endrun(routine//':ERROR: file ilev does not match model.')
    end if
 
    ierr = pio_inq_varid(File, 'hyai', hyai_desc)
+   if (ierr /= PIO_NOERR) then
+      ierr = pio_inq_varid(File, 'sigma_pressure_hybrid_coordinate_a_coefficient_at_interfaces', hyai_desc)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': reading hyai')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90]: Updated!

File: src/dynamics/utils/hycoef.F90 [hunk 4]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
+      end if
+   end if
+
    ierr = pio_inq_varid(File, 'hyam', hyam_desc)
+   if (ierr /= PIO_NOERR) then
+      ierr = pio_inq_varid(File, 'sigma_pressure_hybrid_coordinate_a_coefficient', hyam_desc)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': reading hyam')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90]: Updated!

File: src/dynamics/utils/hycoef.F90 [hunk 5]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
+      end if
+   end if
+
    ierr = pio_inq_varid(File, 'hybi', hybi_desc)
+   if (ierr /= PIO_NOERR) then
+      ierr = pio_inq_varid(File, 'sigma_pressure_hybrid_coordinate_b_coefficient_at_interfaces', hybi_desc)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': reading hybi')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90]: Updated!

File: src/dynamics/utils/hycoef.F90 [hunk 6]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
+      end if
+   end if
+
    ierr = pio_inq_varid(File, 'hybm', hybm_desc)
+   if (ierr /= PIO_NOERR) then
+      ierr = pio_inq_varid(File, 'sigma_pressure_hybrid_coordinate_b_coefficient', hybm_desc)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': reading hybm')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90]: Updated!

File: tools/generate_input_to_stdnames_update.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    print(f"Opening {csv_filepath}")
+    with open(csv_filepath, encoding='ascii') as csvfile:
+        csvdata = csv.reader(csvfile)
+        for row in csvdata:
+            inputname = row[0].split(" ")[0]
+            standardname_match = pattern.fullmatch(row[5].split(" ")[0])
+            if csvdata.line_num < 432 and standardname_match and inputname and "Skipping" not in row[5] and "CCPP" not in row[5]:
+                print(f"Adding {inputname} under {standardname_match.string}")
+                # if standardname_match.string in datamap:
+                #   raise Exception(f"Found duplicate standard name {standardname_match.string} on line {csvdata.line_num}")
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tools/generate_input_to_stdnames_update.py]: Removed!

File: tools/generate_input_to_stdnames_update.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+          <ic_file_input_name>input_name</ic_file_input_name>
+          ...
+        </entry>
+        ...
+      </entries>
+ 
+    """
+    parser = argparse.ArgumentParser(description='')
+    parser.add_argument('--csv-file', type=str, default='CCPP Standard Names - Sheet1.csv', help='')
+    parser.add_argument('--current-map', type=str, default='stdnames_to_inputnames_dictionary.xml', help='')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tools/generate_input_to_stdnames_update.py]: Removed!  This was a hold over from the previous implementation that was not working.

File: tools/inputnames_to_stdnames.py [hunk 9]
```diff
@@ -97,9 +99,12 @@ def parse_command_line(arguments, description):
     parser.add_argument("--stdnames", type=str, required=True,
                         metavar='stdname file',
                         help="Full path to the standard names dictionary (e.g. stdnames_to_inputnames_dictionary.xml)")
+    parser.add_argument('--tphys', type=str, required=True,
+                        metavar='tphysac or tphybs group - REQUIRED',
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tools/inputnames_to_stdnames.py]: Updated!

File: src/dynamics/utils/hycoef.F90 [hunk 10]
```diff
@@ -394,26 +403,68 @@ subroutine hycoef_read(File)
...
+      end if
+   end if
 
    ierr = pio_get_var(File, hyai_desc, hyai)
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90:469]: Added!

File: tools/inputnames_to_stdnames.py [hunk 11]
```diff
@@ -97,9 +99,12 @@ def parse_command_line(arguments, description):
...
+                        help='Group to convert to stdandard names')
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply tools/inputnames_to_stdnames.py]: Good catch!  Was really confused by this one for a sec and then noticed the `not` in front of the string check.

File: Externals_CAM.cfg [hunk 12]
```diff
@@ -2,7 +2,7 @@
 local_path = ccpp_framework
 protocol = git
 repo_url = https://github.com/peverwhee/ccpp-framework
-tag = CPF_0.2.056
+tag = CPF_0.2.057
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply Externals_CAM.cfg]: Updated!

**Review by nusbaume — APPROVED** [review]
Looks great, thanks @mwaxmonsky!

**Review by peverwhee — CHANGES_REQUESTED** [review]
two more tiny error checks!

File: src/dynamics/utils/hycoef.F90 [hunk 13]
```diff
@@ -394,42 +403,95 @@ subroutine hycoef_read(File)
...
    ! PIO traps errors internally, no need to check ierr
 
    ierr = PIO_Inq_DimID(File, 'lev', lev_dimid)
+   if (ierr /= PIO_NOERR) then
+      ierr = PIO_Inq_DimID(File, 'reference_pressure_in_atmosphere_layer', lev_dimid)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': ERROR: unable to find lev dimension in ncdata or restart file.')
+      end if
+   end if
    ierr = PIO_Inq_dimlen(File, lev_dimid, flev)
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90:418]: Added!

File: src/dynamics/utils/hycoef.F90 [hunk 14]
```diff
@@ -394,42 +403,95 @@ subroutine hycoef_read(File)
...
    end if
 
    ierr = PIO_Inq_DimID(File, 'ilev', lev_dimid)
+   if (ierr /= PIO_NOERR) then
+      ierr = PIO_Inq_DimID(File, 'reference_pressure_in_atmosphere_layer_at_interfaces', lev_dimid)
+      if (ierr /= PIO_NOERR) then
+         call endrun(routine//': ERROR: unable to find ilev dimension in ncdata or restart file')
+      end if
+   end if
    ierr = PIO_Inq_dimlen(File, lev_dimid, filev)
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply src/dynamics/utils/hycoef.F90:434]: Added!

**Review by peverwhee — APPROVED** [review]
thanks!

**Comment by peverwhee** [comment]
@mwaxmonsky I ran a clean check-out of CAM-SIMA with your updates and git-fleximod worked well. I did run into this bug - https://github.com/ESCOMP/CAM-SIMA/pull/280 (introduced by me), but I vote for merging this PR!

**Pull request merged**